### ISSUE TITLE
Trying to deserialize the wrong json result in a panic

### ISSUE THREAD

**Issue by gbip** [rust-embedded-community/serde-json-core#11]
When trying to deserialize a valid json that does not represent the expected struct, the deserializer calls [`deserialize_ignored_any`](https://github.com/japaric/serde-json-core/blob/master/src/de/mod.rs#L548) which should be unreachable.

Here is an example inspired by some code that we used in production :
```rust
use serde_json_core::de::{from_slice, Error as DError};

#[derive(Debug, Copy, Clone, Deserialize, Serialize)]
pub enum IOState {
    On,
    Off,
}

#[derive(Debug, Copy, Clone, Deserialize, Serialize)]
pub struct Pneumatic {
    pub pumps: [IOState; 2],
    pub pump_intensity: [u16; 2],
    pub valves: [IOState; 6],
}

impl Pneumatic {
    fn from_json_slice(slice: &[u8]) -> Result<Self, DError> {
        from_slice(slice)
    }
}
```
```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deser_fail() {
        let b = Pneumatic::from_json_slice(
            "{\"pump_intensity\":[0,0],\"pumps\":[\"On\",\"On\"],\"valves\":[\"Off\",\"On\",\"On\",\"Off\",\"On\",\"Off\"]}".as_bytes()).unwrap();
       // This line fails
        let c = Pneumatic::from_json_slice(
            "{\"tirette\":\"Triggered\",\"buzzer\":\"PlayErrorSound\"}".as_bytes(),
        );
    }

}
```

You can find the code to replicate this issue on [this branch](https://github.com/gbip/serde-json-core/tree/panic_replication) (ready to be cloned).

Please tell me if I can do anything else to document this issue ! :smile:

**Comment by fionawhim** [comment]
This threw me as well. 

Would it make sense for `deserialize_ignored_any` to return an `Error` rather than crash?

**Comment by fionawhim** [comment]
I actually was able to implement `deserialize_ignored_any` in #13, which could solve this.

The derive Deserializer implementation uses `deserialize_ignored_any` to toss fields that it doesn’t recognize.

**Comment by webmaster128** [comment]
Is this done with #13 being merged?

**Comment by MathiasKoch** [comment]
Think this issue can be closed with #13 merged?

**Comment by MathiasKoch** [comment]
Closed by #13
